nix-prefetch-scripts: just don't depend on Nix

K900 opened the pull request:
If we're running it from a build (fetchgit, fetchCargoVendor, etc), none of the code paths that actually call Nix are hit.

If we're _not_ running it from a build, the user can reasonably be expected to have a Nix present ambiently.

This avoids pulling Nix into the build closure of fetchCargoVendor, and thus causing infrecs when Nix depends on Rust things.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

TomaSajt commented:
btw `fetchgit` doesn't depend on `nix-prefetch-git`-the-package, but on `nix-prefetch-git`-the-script-file.

`fetchgit` provides its own `git` and `git-lfs` to the build environment the script is run run (but not `nix`).

Personally, I feel like `nix-prefetch-git`-the-package should declare its dependencies...
But I don't care *that* much.

In any case, my other PR would still have to remove `git-lfs`...

K900 commented:
Ideally we'd refactor this into two scripts - nix-fetch-git or whatever for the actual fetching, and then a second wrapper that just does the store path munging. That way tools can depend on the former, and the latter can depend on Nix.

alois31 commented:
Can we also remove the `allowGitDependencies` hack then?

alois31 commented:
So what exactly are we gaining here then? Unless we want to make only git-lfs optional and not the entirety of git.

K900 commented:
The only source of the infrec on git-lfs is Ruby, as far as I can tell. We can probably special case that?

K900 committed:
treewide: drop allowGitDependencies hack

K900 commented:
Some looking later, Cargo does not support git-lfs at all, so stubbing it out is a non-issue, at least for now.

TomaSajt commented:
So: we're dropping nix from the `nix-prefetch-git` script because in reality, no nixpkgs fetcher should ever use the nix part of the script.
And: this is better than the override method because we don't have to do it everywhere.
But: this will reduce the "pureness" of the package.
However: it's not that bad, since we're expected to have nix anyway.

Should the dropping of `allowGitDependencies` be part of this PR?

K900 commented:
> So: we're dropping nix from the nix-prefetch-git script because in reality, no nixpkgs fetcher should ever use the nix part of the script.

Yes.

> And: this is better than the override method because we don't have to do it everywhere.

Also yes.

> But: this will reduce the "pureness" of the package.

I guess so, but ideally the next step would be to split it into two scripts, one that depends on just git (and maybe git-lfs) and another that also depends on Nix. It's not trivial with the way the script is written though, so I feel like we can start here.

> However: it's not that bad, since we're expected to have nix anyway.

For the running it out of nix case, yes.

> Should the dropping of `allowGitDependencies` be part of this PR?

I don't really see why not?

K900 commented on pkgs/build-support/rust/fetch-cargo-vendor.nix:
```diff
@@ -35,9 +35,6 @@ in
   nativeBuildInputs ? [ ],
-  # This is mostly for breaking infinite recursion where dependencies
-  # of nix-prefetch-git use fetchCargoVendor.
-  allowGitDependencies ? true,
```
In the flake? Yes

TomaSajt replied:
It shouldn't blow it up: this attr will just be a NOOP.
(We don't have to have it explicitly be destructured from the function input attrset)

K900 replied:
Oh yeah, true, it takes `...`, so we should be good.

K900 committed:
nix-prefetch-scripts: just don't depend on Nix

If we're running it from a build (fetchgit, fetchCargoVendor, etc),
none of the code paths that actually call Nix are hit.

If we're _not_ running it from a build, the user can reasonably be expected
to have a Nix present ambiently.

This avoids pulling Nix into the build closure of fetchCargoVendor,
and thus causing infrecs when Nix depends on Rust things.

K900 commented:
Do we all agree on this then? Should I send it?

TomaSajt commented:
I still want to check a few things with update scripts and other places `nix-prefetch-git` is used.

K900 commented:
I've looked around and have not found any other in-sandbox invocations of nix-prefetch-git without `--builder`.

TomaSajt commented:
Okay, everything looks OK...
I saw that the other `fetch*` scripts don't even use `nix-prefetch-*`.

IMO there should be a comment in the file stating why nix isn't declared as an input. (For clarity's sake and so that people don't add it back.)

Also small question: do dependency cycles usually go in the direction?
What you wrote:
`nix-prefetch-git -> fetchCargoVendor -> ruby (yjit) -> asciidoctor -> git-lfs -> nix-prefetch-git`
What I think is more readable (going left to right, dependents pointing at dependencies)
`nix-prefetch-git -> git-lfs -> asciidoctor -> ruby (yjit) -> fetchCargoVendor -> nix-prefetch-git`

I wasn't able to find other chains like this in tree (because I guess I'm bad at grepping) so I don't know other examples to see which is usually used.

K900 commented:
I don't think there's a hard rule, but yeah, it makes more sense to me the other way around too. I'll update the comments.

TomaSajt commented on pkgs/tools/package-management/nix-prefetch-scripts/default.nix:
```diff
@@ -66,6 +53,9 @@ let
 rec {
+  # No explicit dependency on Nix, as these can be used inside builders,
+  # and thus will cause dependency loops. When used _outside_ builders,
+  # we expect people to have a Nix implementation available ambiently.
```
Only `nix-prefetch-git` can be used.
Though we can keep the comment in this place, it's fine.

TomaSajt reviewed (approved):
LGTM

Hopefully we don't break anything.
